### PR TITLE
[2646] Avoid rerendering the palette if not necessary

### LINKED ISSUE #2646: The palette rerenders too much
### Steps to reproduce
Log something in the DiagramElementPalette or DiagramPalette composant.

### Expected behavior

When the palette is closed, we should not see any logs.

### Actual behavior

Since we're using the useViewport(); hook, the palette re renders when the viewport changes, meaning the palette gets re rendered a lot during a move.

### What to do

We should get the value of the viewport directly from the store instead of using the useViewport() hook.
```
  const store = useStoreApi();
 const [viewportX, viewportY, viewportZoom] = [...transform];
```

### PR DESCRIPTION
**Pull request by mcharfadi**
Bug: https://github.com/eclipse-sirius/sirius-web/issues/2646

# Pull request template

## General purpose
What is the main goal of this pull request?
- [x] Bug fixes
- [ ] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/palette/DiagramElementPalettePortal.tsx [hunk 1]
```diff
@@ -12,12 +12,12 @@
...
-import { ReactFlowState, useStore } from 'reactflow';
-
-const selector = (state: ReactFlowState) => state.domNode?.querySelector('.react-flow__renderer');
+import { useStoreApi } from 'reactflow';
 
 export const DiagramElementPalettePortal = ({ children }: { children: ReactNode }) => {
-  const wrapperRef = useStore(selector);
+  const store = useStoreApi();
+  const { domNode } = store.getState();
+  const wrapperRef = domNode?.querySelector('.react-flow__renderer');
```
Comment (sbegaudeau) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/palette/DiagramElementPalettePortal.tsx]: What is the point of this change?

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/palette/DiagramPalettePortal.tsx [hunk 2]
```diff
@@ -12,13 +12,13 @@
...
+import { useStoreApi } from 'reactflow';
 import { DiagramPalettePortalProps } from './DiagramPalette.types';
 
-const selector = (state: ReactFlowState) => state.domNode?.querySelector('.react-flow__renderer');
-
 export const DiagramPalettePortal = ({ children }: DiagramPalettePortalProps) => {
-  const wrapperRef = useStore(selector);
+  const store = useStoreApi();
+  const { domNode } = store.getState();
+  const wrapperRef = domNode?.querySelector('.react-flow__renderer');
```
Comment (sbegaudeau) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/palette/DiagramPalettePortal.tsx]: What is the point of this change?

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/palette/usePaletteReferencePosition.tsx [hunk 3]
```diff
@@ -10,13 +10,16 @@
...
+import { useStoreApi } from 'reactflow';
 import { useDiagramElementPalette } from './useDiagramElementPalette';
 import { useDiagramPalette } from './useDiagramPalette';
+import { UsePaletteReferencePositionValue } from './usePaletteReferencePosition.types';
 
 export const usePaletteReferencePosition = (): UsePaletteReferencePositionValue => {
-  const { x: viewportX, y: viewportY, zoom: viewportZoom } = useViewport();
+  const store = useStoreApi();
+  const { transform } = store.getState();
+  const [viewportX, viewportY, viewportZoom] = [...transform];
```
Comment (sbegaudeau) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/palette/usePaletteReferencePosition.tsx]: What is the point of this change?

**Review by sbegaudeau — CHANGES_REQUESTED** [review]
I don't see why most of the changes are being done

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/palette/usePaletteReferencePosition.tsx (hunk 3 above)
Comment (mcharfadi) [reply packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/palette/usePaletteReferencePosition.tsx]: When using useViewport() hook, the component will re render each time the viewport change. 
With this change we avoid unnecessary rendering.

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/palette/DiagramElementPalettePortal.tsx (hunk 1 above)
Comment (mcharfadi) [reply packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/palette/DiagramElementPalettePortal.tsx]: When using useStore() we subscribe to the state and the component will re render each time the state changes. 
With useStoreApi() we get the value when needed.
With this change we avoid unnecessary rendering.

**Comment by sbegaudeau** [comment]
It is quite easy to stop the unnecessary rendering of the palette but most of the issues are not coming from here but instead from the handles and how they are refreshed.

**Review by sbegaudeau — APPROVED** [review]

**Pull request merged**